### DISCUSSION TITLE
I create a hook for autogenerating an avatar using the improved 0.10 upsert form

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1418] [Show and tell]
In 0.10 the RecordUpsert form received some new methods for local uploading.  The only trick is to create assign a new Daos struct that doesn't have the record hooks on it.  This prevents an infinite loop, or other tricks to break out of it.

```golang
	app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {
		m, ok := e.Model.(*pbmodels.Record)
		if !ok {
			return nil
		}

		switch m.TableName() {
		case models.UsersCollectionName:
			return user_hooks.AfterCreate(env, m)
		}

		return nil
	})

```

```golang
func AfterCreate(env *svc.Env, m *pbmodels.Record) error {
	if m.GetString("avatar") == "" {
		return generateAvatar(env, m)
	}

	return nil
}

func generateAvatar(env *svc.Env, m *pbmodels.Record) error {
	g := govatar.MALE
	p := "/tmp/gravatar.jpg"

	if err := avatar.GenerateForName(g, m.GetString("username"), models.UsersAvatarSize, p); err != nil {
		return err
	}

	f, err := filesystem.NewFileFromPath(p)
	if err != nil {
		return err
	}

	form := forms.NewRecordUpsert(env.App, m)
	form.SetDao(daos.New(env.Dao().DB()))
	form.AddFiles("avatar", f)

	return form.Submit()
}


```

**Comment by dschissler** [comment]
@ganigeorgiev Do you know how to get a `ReadSeekCloser` from an image struct or buffer without saving a file to disk?  It would be nice to do it all in memory.

**Reply by ganigeorgiev** [reply]
I'm not sure. If you have raw bytes and want to implement the `ReadSeekCloser` interface I guess the easiest way would be to create a wrapper struct that embeds the std `bytes.Reader` with empty? `Close`, eg. something like:

```go
type BytesReadSeekCloser struct {
	*bytes.Reader
}

func NewBytesReadSeekCloser(b []byte) *BytesReadSeekCloser {
	return &BytesReadSeekCloser{bytes.NewReader(b)}
}

func (r *BytesReadSeekCloser) Close() error {
        // eventually the `Close` could reset the seeker...
	return nil
}
```
(_I haven't tested the above and it is possible that there is already an existing structure in the std lib that do what you want_)

**Reply by dschissler** [reply]
@ganigeorgiev Could you add a `NewFileFromBuf` or `NewFileFromBytes` helper method?  It would really help out with real life situations.  In my `GenerateForName` function I'm call [govatar](https://github.com/o1egl/govatar) which returns a byte slice.  I'm only saving it because that is what the Pocketbase requires.  It would be so much easier if I could just shove the bytes using `filesystem.NewFileFromBytes(buf)`.  That will save from doing so much crap in hooks.

**Comment by dschissler** [comment]
This is the updated version with the new `filesystem.NewFileFromBytes` method that allows me to skip the step of writing to an OS file (on disk).

```golang
	app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {
		m, ok := e.Model.(*pbmodels.Record)
		if !ok {
			return nil
		}

		switch m.TableName() {
		case models.UsersCollectionName:
			return user_hooks.AfterCreate(env, m)
		}

		return nil
	})

```

```golang
func AfterCreate(env *svc.Env, m *pbmodels.Record) error {
	if m.GetString("avatar") == "" {
		return generateAvatar(env, m)
	}

	return nil
}

func generateAvatar(env *svc.Env, m *pbmodels.Record) error {
	g := govatar.MALE

	b, err := avatar.GenerateForName(g, m.GetString("username"), models.UsersAvatarSize)
	if err != nil {
		return err
	}

	f, err := filesystem.NewFileFromBytes(b, "avatar.jpg")
	if err != nil {
		return errors.New("unable to create avatar file from bytes")
	}

	form := forms.NewRecordUpsert(env.App, m)
	form.SetDao(daos.New(env.Dao().DB()))
	form.AddFiles("avatar", f)

	return form.Submit()
}
```
